src/java.base/share/classes/sun/security/rsa/RSASignature.java

```diff
@@ -205,24 +208,29 @@ protected boolean engineVerify(byte[] sigBytes) throws SignatureException {
         }
         try {
             if (sigBytes.length != RSACore.getByteLength(publicKey)) {
-                throw new SignatureException("Signature length not correct: got " +
+                throw new SignatureException("Bad signature length: got " +
                     sigBytes.length + " but was expecting " +
                     RSACore.getByteLength(publicKey));
             }
+
+            // https://www.rfc-editor.org/rfc/rfc8017.html#section-8.2.2
+            // Step 4 suggests comparing the encoded message
             byte[] digest = getDigestValue();
+            byte[] encoded = RSAUtil.encodeSignature(digestOID, digest);
+            byte[] padded = padding.pad(encoded);
             byte[] decrypted = RSACore.rsa(sigBytes, publicKey);
-            byte[] unpadded = padding.unpad(decrypted);
-            // https://www.rfc-editor.org/rfc/rfc8017.html#section-8.2.2
-            // Step 4 suggests comparing the encoded message instead of the
-            // decoded, but some vendors might omit the NULL params in
-            // digest algorithm identifier.
-            byte[] decodedDigest = RSAUtil.decodeSignature(digestOID, unpadded);
-            return MessageDigest.isEqual(digest, decodedDigest);
+
+            boolean status = MessageDigest.isEqual(padded, decrypted);
+            if (!status) {
+                // fall back to the decode approach for max compatibility
+                byte[] unpadded = padding.unpad(decrypted);
+                if (unpadded != null) {
+                    status = MessageDigest.isEqual(digest,
+                            RSAUtil.decodeSignature(digestOID, unpadded));
+                }
+            }
```

Comment:
I understand where the fallback code came from.  As the padding code is exactly the same as engineSign(), the risk may be minimal.  With the fallback code, the security concern (time-constant) we cared about will come back.  Did you run into testing failure without the fallback doe?

Reply:
@XueleiFan No failure observed in regression tests as well as Max's preliminary interop testing. However, it is suggested to ensure max compatibility as you know.

Are you ok with @ferakocz 's suggestion? I can switch to that approach if no objection.

Reply:
I'm OK with @ferakocz's suggestion, by avoiding the use of null params.

Reply:
@ferakocz So, with this approach, we are paying the extra cost of encode signature + pad (for the omit null case) even for impls conforming to RFC 8017 spec. Based on the current interoperability testing, do you still feel that this is worthwhile to do?

Reply:
Well, for conforming implementations we just do the first check and succeed. What I suggested was that we do the encode without null params and pad() *instead* of the fallback to unpad()decodeSignature(). As I said, this part need not be constant time (except for the byte array comparison part), but it can even be made constant time to satisfy the purists :-) at the expense of an extra encode/pad operation which is not that expensive.

Reply:
If it need not be constant time, I'd prefer to not pay the extra cost for every operation and for all callers.